homeassistant/components/pi_hole/__init__.py

```diff
@@ -215,11 +148,52 @@ async def do_enable(name):
         DOMAIN, SERVICE_ENABLE, enable_service_handler, schema=service_enable_schema
     )
 
-    hass.async_create_task(async_load_platform(hass, SENSOR_DOMAIN, DOMAIN, {}, config))
+    return True
+
+
+async def async_setup_entry(hass, entry):
+    """Set up Pi-hole entry."""
+    name = entry.data[CONF_NAME]
+    host = entry.data[CONF_HOST]
+    use_tls = entry.data[CONF_SSL]
+    verify_tls = entry.data[CONF_VERIFY_SSL]
+    location = entry.data[CONF_LOCATION]
+    api_key = entry.data.get(CONF_API_KEY)
+
+    LOGGER.debug("Setting up %s integration with host %s", DOMAIN, host)
+
+    try:
+        session = async_get_clientsession(hass, verify_tls)
+        pi_hole = PiHoleData(
+            Hole(
+                host,
+                hass.loop,
+                session,
+                location=location,
+                tls=use_tls,
+                api_token=api_key,
+            ),
+            name,
+        )
+        await pi_hole.async_update()
+        hass.data[DOMAIN][name] = pi_hole
```

Comment:
It's safer and standard to use the entry.entry_id as key when storing entry related data in hass.data. The `name` is user configurable.

Reply:
Yeah I'm planning on change the way the service calls work and maybe I would be able to remove that completely.